android/maplibreui/src/main/java/com/stadiamaps/ferrostar/maplibreui/config/MapControls.kt

```diff
@@ -1,34 +1,97 @@
- * @param isArrivalExpanded weather the arrival view is expanded.
+ * @param arrivalViewHeight The height of the arrival view.
+ * @param horizontalPadding The horizontal padding to apply to the map controls. Defaults to 16.dp
+ *   to match the default padding of the arrival view.
+ * @param verticalPadding The vertical padding to apply to the map controls from the top of the
+ *   arrival view. Defaults to 8.dp.
+ *
+ * TODO: This function is attempting to optimize the map controls for many screen sizes, system
+ *   bars, and orientations. We should remain open to feedback for specific cases.
  */
-internal fun mapControlsFor(isLandscape: Boolean, isArrivalExpanded: Boolean): MapControls {
-  val bottom =
-      when {
-        isLandscape -> 32
-        isArrivalExpanded -> 32 * 9
-        else -> 32 * 7
-      }
-
-  return MapControls(
-      attribution =
-          AttributionSettings(
-              gravity = Gravity.BOTTOM or Gravity.END,
-              margins = MarginInsets(end = 270, bottom = bottom)),
-      compass = CompassSettings(enabled = false),
-      logo =
-          LogoSettings(
-              gravity = Gravity.BOTTOM or Gravity.END,
-              margins = MarginInsets(end = 32, bottom = bottom)))
+@Composable
+internal fun rememberMapControlsFor(
+    arrivalViewHeight: Dp = 0.dp,
+    horizontalPadding: Dp = 16.dp,
+    verticalPadding: Dp = 8.dp
+): State<MapControls> {
+  val rawLayoutDirection = LocalConfiguration.current.layoutDirection
+  val layoutOrientation: LayoutDirection = LayoutDirection.entries[rawLayoutDirection]
+
+  val isLandscape = LocalConfiguration.current.orientation == Configuration.ORIENTATION_LANDSCAPE
+
+  val gridPadding = paddingForGridView()
+  val windowInsetPadding = WindowInsets.systemBars.asPaddingValues()
+
+  // This calculation clamps the controls to the trailing edge of the screen in landscape mode
+  // with less padding in that case. The reason for this is that with edge-to-edge, there's
+  // a larger map canvas available.
+  val endPaddingDp =
+      windowInsetPadding.calculateEndPadding(layoutOrientation) +
+          gridPadding.calculateEndPadding(layoutOrientation)
+  val endOffsetDp = if (isLandscape) endPaddingDp else horizontalPadding
+
+  val bottomPaddingDp =
+      windowInsetPadding.calculateBottomPadding() + gridPadding.calculateBottomPadding()
+  val bottomOffsetDp = if (isLandscape) bottomPaddingDp else bottomPaddingDp + arrivalViewHeight
+
+  // TODO: This could be improved if we want to add pixel width to dp conversion in
+  //  maplibre-compose.
+  val attributionOffset = 24.dp
+
+  val layoutDirection = localLayoutDirection()
+  val density = LocalDensity.current
+
+  // Create a mutable state for the map controls. This will be updated in the LaunchedEffect
+  // whenever the layout input values change
+  val mapControlsState = rememberSaveable { mutableStateOf(MapControls()) }
+
+  LaunchedEffect(arrivalViewHeight, gridPadding, windowInsetPadding) {
```

Comment:
I'm not entirely convinced that a launched effect is the "correct" way to accomplish this. It sounds like the point of this IS to do side effects. But it's very specifically framed around suspend functions in the docs. The point of this isn't actually to do any sort of suspending operation at all but rather to do something on-off when some state changes. That feels more like "normal" Composable state to me, but I can kinda see the temptation to use this. I suspect this is at least partially to blame though.

Reply:
This discussion https://developer.android.com/develop/ui/compose/side-effects points me toward `produceState`. 

> Even though produceState creates a coroutine, it can also be used to observe non-suspending sources of data. To remove the subscription to that source, use the [awaitDispose](https://developer.android.com/reference/kotlin/androidx/compose/runtime/ProduceStateScope#awaitDispose(kotlin.Function0)) function.

It seems to tidily contain the calculations. That said, since it's still running a coroutine scope, it also causes the crashy/buggy behavior.

I'd say we use it as we continue to debug the crash data you isolated?

Reply:
Good find; this looks more fit to the task to me!